### PR TITLE
Support for Kubernetes v1.20

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area open-source usability
/kind enhancement
/priority normal
/merge keep-commits
/exp intermediate
/topology garden seed shoot

**What this PR does / why we need it**:
Support for Kubernetes v1.20

**Which issue(s) this PR fixes**:
Part of #3260 

**Special notes for your reviewer**:
* I didn't vendor new `k8s.io/*` Golang sources yet - @timebertt plans to update to `kubernetes-v1.19` as part of #3109.
* ⚠️ This PR only contains the Gardener part for supporting 1.20 - there will/must be follow-up PRs for each extension repository where individual support gets added.

**Release note**:

```feature developer
Gardener can now support shoot clusters with Kubernetes version 1.20. In order to allow creation/update of 1.20 clusters you will have to update the version of your provider extension(s) to a version that supports 1.20 as well. Please consult the respective releases and notes in the provider extension's repository.
```

### CODE REVIEW

File: charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ spec:
     kind: Hvpa
     plural: hvpas
   scope: ""
+  preserveUnknownFields: false
```
Comment (ialidzhikov) [inline charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml:14]: Is this working for Seed clusters that are `< v1.15`? To my knowledge `preserveUnknownFields` is available only for `>= v1.15`.

File: pkg/utils/kubernetes/leaderelection.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"context"
+	fmt "fmt"
```
Comment (ialidzhikov) [inline pkg/utils/kubernetes/leaderelection.go]:
```suggestion
	"fmt"
```

**Review by ialidzhikov — COMMENTED** [review]
Thank you for the PR! I still need to go through in details and play with it.

File: charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml (hunk 1 above)
Comment (rfranzke) [reply charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml:14]: Nice catch, I've added some `if-end` statements. Though, I guess in earlier Kubernetes versions additional fields in the sent JSONs are ignored. Thanks for finding!

**Comment by ialidzhikov** [comment]
Apart from the failed verify step, it looks very good to me. I played with the PR locally extensively (seed bootstrap for v1.20; hibernation and wake up with v1.20; dynamic volume provisioning with CSI; worker pool with CRI; switch between IPVS and IPTables; worker with suse-chost) and it worked as expected.

/lgtm

File: Makefile [hunk 3]
```diff
@@ -191,6 +192,10 @@ check:
 generate:
 	@hack/generate.sh ./cmd/... ./extensions/... ./pkg/... ./plugin/... ./test/...
 
+.PHONY: generate-extensions-crds
+generate-extensions-crds:
+	@controller-gen crd paths=./pkg/apis/extensions/... output:crd:dir=./dev/extensions-crds output:stdout
```
Comment (timebertt) [inline Makefile:197]: Nice, also wanted to propose using controller-tools for that :) 
We could also generate the CRDs in `charts/seed-bootstrap/templates/extensions` again, right?
Should we move this to `hack/update-codegen.sh` or use a `go:generate` comment in `pkg/apis/extensions`?

Comment (rfranzke) [reply Makefile:197]: > We could also generate the CRDs in charts/seed-bootstrap/templates/extensions again, right?

WDYM? I generated them and added/updated the open API validation spec. Did I forget something?

> Should we move this to hack/update-codegen.sh or use a go:generate comment in pkg/apis/extensions?

This won't be easily possible (we can't just blindly take what was generated from this CLI tool), especially now that we have `if-end` expressions in the CRDs.

**Comment by rfranzke** [comment]
@ialidzhikov I've pushed a new commit 588143d which fixes the `verify` step, PTAL.

> I played with the PR locally extensively (seed bootstrap for v1.20; hibernation and wake up with v1.20; dynamic volume provisioning with CSI; worker pool with CRI; switch between IPVS and IPTables; worker with suse-chost) and it worked as expected.

That's awesome, thank you very much for your time and efforts and extensive testing, highly appreciated. 👏

File: Makefile (hunk 3 above)
Comment (timebertt) [reply Makefile:197]: > This won't be easily possible (we can't just blindly take what was generated from this CLI tool), especially now that we have if-end expressions in the CRDs.

Yeah, sorry. Forgot about the `if`s in the CRDs.
Then you can ignore my comment.

Comment (rfranzke) [reply Makefile:197]: OK, thanks!

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by timuthy — APPROVED** [review]
/lgtm
Nice job! 👏

**Comment by rfranzke** [comment]
Thanks for your reviews, guys! I'm going ahead merging this now and marking the respective PRs on the provider extensions as ready for review.

**Pull request merged**